Rewrite Grpc.Tools BUILD-INTEGRATION.md docs

tonydnewell opened the pull request:
Major rewrite and refactoring of the Grpc.Tools documenation. Viewing a diff may not help as lots has been moved around or rewritten.

I've removed the chatty style, added an introduction, moved (and rewritten) the examples to after the reference section.

tonydnewell commented:
@jtattermusch for review

jtattermusch commented on src/csharp/BUILD-INTEGRATION.md:
```diff
@@ -394,13 +136,13 @@ otherwise unsupported (e.g. experimental) flags to protocol buffer compiler.
 
 * __GrpcOutputOptions__ 
```
inconsistency: the Notes section mentions GrpcOutputOptions but doesn't mention OutputOptions - please add it.

jtattermusch commented on src/csharp/BUILD-INTEGRATION.md:
```diff
@@ -440,3 +182,256 @@ protoc --plugin=protoc-gen-grpc=grpc_csharp_plugin --csharp_out=OUT_DIR \
+E.g. the directory `"...\.nuget\packages\grpc.tools\2.51.0\build\native\include"` is automatically included.  This is the directory where a declaration such as `import "google/protobuf/wrappers.proto";` in a proto file would be found.
+
+---
+## <a name="ProtoRoot"></a>ProtoRoot - Avoiding duplicate generated file names
```
AFAIK the ProtoRoot is especially important for making sure that the  "import foo/bar/b.proto" statements will work as expected.  Basically, if you have `Protos/foo/bar/a.proto` and `Protos/foo/bar/b.proto` and a.proto imports b.proto via `import foo/bar/b.proto`, the ProtoRoot needs to be set to "Protos" otherwise the protoc won't see b.proto and build will fail.
This isn't something that's mentioned in the doc currently, but it's very important and people often get it wrong.

jtattermusch replied:
The example title talks about "Avoiding duplicate generate file names" but the contents of this section don't match that. There no duplicate file name in `<Protobuf Include="**\*.proto" />` nor `<Protobuf Include="**\*.proto" ProtoRoot="Protos" />`.
But if there would be another .proto file that does `import v1/hello.proto`, the example would work without also setting ProtoRoot.

jtattermusch commented on src/csharp/BUILD-INTEGRATION.md:
```diff
@@ -1,351 +1,93 @@
-below stands for a command prompt in either platform):
-
+<ItemGroup>
+    <Protobuf Include="**\*.proto" />
```
This doc doesn't mention the option to enable auto-inclusion of .proto files anywhere.
See https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Tools/implementation_notes.md#automatically-including-proto-files  - perhaps worth mentioning this somewhere?

jtattermusch commented:
I created a PR with a few typo fixes and other stuff against this change's branch: https://github.com/tonydnewell/grpc/pull/1

tonydnewell commented:
@jtattermusch I've merged in your changes and taken into account the review comments. I've also rewritten the sections on ProtoRoot and AdditionalImportDirs, added information about auto including .proto files, and added more properties.

Ready for further review.

tonydnewell commented:
@jtattermusch Also added README.md to the package

jtattermusch commented on src/csharp/Grpc.Tools/Grpc.Tools.csproj:
```diff
@@ -35,20 +35,24 @@
     <Authors>The gRPC Authors</Authors>
     <Copyright>Copyright 2018 The gRPC Authors</Copyright>
-    <Description>gRPC and Protocol Buffer compiler for managed C# and native C++ projects.
+    <Description>gRPC and Protocol Buffer compiler for C# projects.
```
I checked other packages on nuget and when they provide a README.md, their descriptions is usually a single line with the very basic info - let's update the description here to match that.

![image](https://user-images.githubusercontent.com/9939684/214001123-c9075136-b758-47bc-8943-c1118d86ad17.png)

jtattermusch commented on src/csharp/BUILD-INTEGRATION.md:
```diff
@@ -28,11 +14,17 @@ The package [Grpc.Tools](https://www.nuget.org/packages/Grpc.Tools) is used to g
 
-It isn't required at runtime, so the dependency should be marked with `PrivateAssets="All"`, e.g.
+`Grpc.Tools` is only used at build-time and has no runtime components.
+It should be marked with `PrivateAssets="All"` to exclude it from being included at runtime, e.g.
```
nit: "to exclude it from being included at runtime" sounds odd. rephrase?

jtattermusch commented on src/csharp/BUILD-INTEGRATION.md:
````diff
@@ -222,67 +220,82 @@ For other examples see also the `.csproj` files in the examples in GitHub:
-```
-	ProjectFolder\Protos\v1\hello.proto
-	ProjectFolder\Protos\v2\hello.proto
+By default the output directory for generated files is `obj\RELEASE\FRAMEWORK\` (e.g. `obj\Debug\net6.0\`) unless `OutputDir` or `GrpcOutputDir` are specified.
````
nit: I think "RELEASE" is called CONFIGURATION in the msbuild lingo?

jtattermusch reviewed (approved):
LGTM, with a few more nits.

tonydnewell commented:
@jtattermusch I've addressed the review comments.
